File: crates/globstar/tests/compiler_stages.rs
use std::fmt::Write as _;

use globstar::ast::{ClassItem, Node};
use globstar::engine::ops::{Op, lower, lower_owned};
use globstar::parser::parse;

const CASES: &str = include_str!("../../../fixtures/compiler-stages.tsv");

fn bytes(out: &mut String, value: &[u8]) {
    for &b in value {
        match b {
            b'\\' => out.push_str("\\\\"),
            b'(' | b')' | b',' | b'|' | b'[' | b']' => {
                let _ = write!(out, "\\x{b:02x}");
            }
            0x20..=0x7e => out.push(b as char),
            _ => {
                let _ = write!(out, "\\x{b:02x}");
            }
        }
    }
}

fn ast_dump(node: &Node, out: &mut String) {
    match node {
        Node::Concat(children) => {
            out.push_str("C(");
            for (i, child) in children.iter().enumerate() {
                if i != 0 {
                    out.push(',');
                }
                ast_dump(child, out);
            }
            out.push(')');
        }
        Node::Literal(value) => {
            out.push_str("L(");
            bytes(out, value);
            out.push(')');
        }
        Node::Separator => out.push('S'),
        Node::AnyChar => out.push('Q'),
        Node::Star => out.push('T'),
        Node::Globstar => out.push('G'),
        Node::Class(class) => {
            out.push_str(if class.negated { "N(" } else { "K(" });
            for (i, item) in class.items.iter().enumerate() {
                if i != 0 {
                    out.push(',');
                }
                match item {
                    ClassItem::Byte(b) => bytes(out, &[*b]),
                    ClassItem::Range(lo, hi) => {
                        bytes(out, &[*lo]);
                        out.push('-');
                        bytes(out, &[*hi]);
                    }
                }
            }
            out.push(')');
        }
        Node::Brace(branches) => {
            out.push_str("B(");
            for (i, branch) in branches.iter().enumerate() {
                if i != 0 {
                    out.push('|');
                }
                ast_dump(branch, out);
            }
            out.push(')');
        }
    }
}

fn ops_dump(ops: &[Op], out: &mut String) {
    for (i, op) in ops.iter().enumerate() {
        if i != 0 {
            out.push(',');
        }
        match op {
            Op::Lit(value) => {
                out.push_str("L(");
                bytes(out, value);
                out.push(')');
            }
            Op::AnyChar => out.push('Q'),
            Op::Star => out.push('T'),
            Op::Class(_) => out.push('K'),
            Op::Sep => out.push('S'),
            Op::SepRun => out.push('R'),
            Op::Globstar => out.push('G'),
            Op::OptSegmentsSlash => out.push('O'),
            Op::SlashAnything => out.push('A'),
            Op::GlobstarAny => out.push('Y'),
            Op::LeadingSeps => out.push('H'),
            Op::Alternation(branches) => {
                out.push_str("B(");
                for (j, branch) in branches.iter().enumerate() {
                    if j != 0 {
                        out.push('|');
                    }
                    out.push('[');
                    ops_dump(branch, out);
                    out.push(']');
                }
                out.push(')');
            }
        }
    }
}

#[test]
fn parser_and_lowering_match_shared_golden_cases() {
    for (line_no, line) in CASES.lines().enumerate() {
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let [pattern, expected_ast, expected_ops] = line
            .split('\t')
            .collect::<Vec<_>>()
            .try_into()
            .unwrap_or_else(|_| panic!("bad fixture line {}", line_no + 1));
        let parsed = parse(pattern.as_bytes()).expect("fixture pattern parses");
        let mut actual_ast = String::new();
        ast_dump(&parsed.body, &mut actual_ast);
        assert_eq!(actual_ast, expected_ast, "AST for {pattern:?}");

        let program = lower(&parsed.body, false);
        let mut actual_ops = String::new();
        ops_dump(program.ops(), &mut actual_ops);
        assert_eq!(actual_ops, expected_ops, "ops for {pattern:?}");

        let owned = lower_owned(parsed.body, false);
        let mut owned_ops = String::new();
        ops_dump(owned.ops(), &mut owned_ops);
        assert_eq!(owned_ops, expected_ops, "owned ops for {pattern:?}");
    }
}

File: packages/globstar/tests/compiler-stages.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";

import {
  CI_BYTE,
  N_ANYCHAR,
  N_BRACE,
  N_CLASS,
  N_CONCAT,
  N_GLOBSTAR,
  N_LITERAL,
  N_SEPARATOR,
  N_STAR,
} from "../src/ast.js";
import { parse } from "../src/parser.js";
import {
  OP_ALTERNATION,
  OP_ANYCHAR,
  OP_CLASS,
  OP_GLOBSTAR,
  OP_GLOBSTAR_ANY,
  OP_LEADING_SEPS,
  OP_LIT,
  OP_OPT_SEGMENTS_SLASH,
  OP_SEP,
  OP_SEP_RUN,
  OP_SLASH_ANYTHING,
  OP_STAR,
  lower,
} from "../src/engine/ops.js";
import { assertNormalizedProgram } from "../src/engine/ops/ir.js";
import { dedupePrefixes } from "../src/engine/ops/prefixes.js";

const here = dirname(fileURLToPath(import.meta.url));
const cases = readFileSync(resolve(here, "../../../fixtures/compiler-stages.tsv"), "utf8");

function bytes(value) {
  let out = "";
  for (const b of value) {
    if (b === 0x5c) out += "\\\\";
    else if (b === 0x28 || b === 0x29 || b === 0x2c || b === 0x7c || b === 0x5b || b === 0x5d)
      out += `\\x${b.toString(16).padStart(2, "0")}`;
    else if (b >= 0x20 && b <= 0x7e) out += String.fromCharCode(b);
    else out += `\\x${b.toString(16).padStart(2, "0")}`;
  }
  return out;
}

function astDump(node) {
  switch (node.tag) {
    case N_CONCAT:
      return `C(${node.children.map(astDump).join(",")})`;
    case N_LITERAL:
      return `L(${bytes(node.bytes)})`;
    case N_SEPARATOR:
      return "S";
    case N_ANYCHAR:
      return "Q";
    case N_STAR:
      return "T";
    case N_GLOBSTAR:
      return "G";
    case N_CLASS:
      return `${node.neg ? "N" : "K"}(${node.items
        .map((item) =>
          item.tag === CI_BYTE ? bytes([item.b]) : `${bytes([item.lo])}-${bytes([item.hi])}`,
        )
        .join(",")})`;
    case N_BRACE:
      return `B(${node.branches.map(astDump).join("|")})`;
    default:
      throw new Error(`unknown AST node ${node.tag}`);
  }
}

function opsDump(ops) {
  return ops
    .map((op) => {
      switch (op.kind) {
        case OP_LIT:
          return `L(${bytes(op.bytes)})`;
        case OP_ANYCHAR:
          return "Q";
        case OP_STAR:
          return "T";
        case OP_CLASS:
          return "K";
        case OP_SEP:
          return "S";
        case OP_SEP_RUN:
          return "R";
        case OP_GLOBSTAR:
          return "G";
        case OP_OPT_SEGMENTS_SLASH:
          return "O";
        case OP_SLASH_ANYTHING:
          return "A";
        case OP_GLOBSTAR_ANY:
          return "Y";
        case OP_LEADING_SEPS:
          return "H";
        case OP_ALTERNATION:
          return `B(${op.branches.map((branch) => `[${opsDump(branch)}]`).join("|")})`;
        default:
          throw new Error(`unknown op ${op.kind}`);
      }
    })
    .join(",");
}

for (const [index, line] of cases.split("\n").entries()) {
  if (!line || line.startsWith("#")) continue;
  const columns = line.split("\t");
  assert.equal(columns.length, 3, `fixture line ${index + 1}`);
  const [pattern, expectedAst, expectedOps] = columns;
  const parsed = parse(pattern);
  assert.equal(astDump(parsed.body), expectedAst, `AST for ${JSON.stringify(pattern)}`);
  const program = lower(parsed.body, false);
  assertNormalizedProgram(program.ops);
  assert.equal(opsDump(program.ops), expectedOps, `ops for ${JSON.stringify(pattern)}`);
}

console.log("✓ shared parser/lowering golden cases");

const encoded = ["src", "src/cli", "src-other", "src2", "src"].map((s) =>
  new TextEncoder().encode(s),
);
assert.deepEqual(
  dedupePrefixes(encoded).map((p) => new TextDecoder().decode(p)),
  ["src", "src2", "src-other"],
);
console.log("✓ prefix dedupe directory boundaries");
